server/fleet/integrations.go

```diff
@@ -0,0 +1,427 @@
+	Email                 string `json:"email"`
+	APIToken              string `json:"api_token"`
+	GroupID               int64  `json:"group_id"`
+	EnableFailingPolicies bool   `json:"enable_failing_policies"`
+}
+
+// ToZendeskIntegration converts a TeamZendeskIntegration to a ZendeskIntegration
+// struct, leaving additional fields to their zero value.
+func (ti TeamZendeskIntegration) ToZendeskIntegration() *ZendeskIntegration {
+	return &ZendeskIntegration{TeamZendeskIntegration: ti}
+}
+
+// JiraIntegration configures an instance of an integration with the Jira
+// system.
+type JiraIntegration struct {
+	// It is a superset of TeamJiraIntegration.
+	TeamJiraIntegration
+
+	EnableSoftwareVulnerabilities bool `json:"enable_software_vulnerabilities"`
+}
+
+// IndexJiraIntegrations indexes the provided Jira integrations in a map keyed
+// by the project key. It returns an error if a duplicate configuration is
+// found for the same project key. This is typically used to index the original
+// integrations before applying the changes requested to modify the AppConfig.
+//
+// Note that the returned map uses non-pointer JiraIntegration struct values,
+// so that any changes to the original value does not modify the value in the
+// map. This is important because of how changes are merged with the original
+// AppConfig when modifying it.
+func IndexJiraIntegrations(jiraIntgs []*JiraIntegration) (map[string]JiraIntegration, error) {
+	byProjKey := make(map[string]JiraIntegration, len(jiraIntgs))
+	for _, intg := range jiraIntgs {
+		if _, ok := byProjKey[intg.ProjectKey]; ok {
+			return nil, fmt.Errorf("duplicate Jira integration for project key %s", intg.ProjectKey)
+		}
+		byProjKey[intg.ProjectKey] = *intg
+	}
+	return byProjKey, nil
+}
+
+// IndexTeamJiraIntegrations is the same as IndexJiraIntegrations, but for
+// team-specific integration structs.
+func IndexTeamJiraIntegrations(teamJiraIntgs []*TeamJiraIntegration) (map[string]TeamJiraIntegration, error) {
+	jiraIntgs := make([]*JiraIntegration, len(teamJiraIntgs))
+	for i, t := range teamJiraIntgs {
+		jiraIntgs[i] = t.ToJiraIntegration()
+	}
+
+	indexed, err := IndexJiraIntegrations(jiraIntgs)
+	if err != nil {
+		return nil, err
+	}
+
+	teamIndexed := make(map[string]TeamJiraIntegration, len(indexed))
+	for k, v := range indexed {
+		teamIndexed[k] = v.TeamJiraIntegration
+	}
+	return teamIndexed, nil
+}
+
+// ValidateJiraIntegrations validates that the merge of the original and new
+// Jira integrations does not result in any duplicate configuration, and that
+// each modified or added integration can successfully connect to the external
+// Jira service.
+//
+// On successful return, the newJiraIntgs slice is ready to be saved - it may
+// have been updated using the original integrations if the API token was
+// missing.
+func ValidateJiraIntegrations(ctx context.Context, oriJiraIntgsByProjKey map[string]JiraIntegration, newJiraIntgs []*JiraIntegration) error {
```

Comment:
Note that all those validation calls are significantly tested via the service's integrations tests.